core/src/main/java/org/infinispan/util/concurrent/locks/impl/DefaultLockManager.java

```diff
@@ -0,0 +1,395 @@
+
+      private CompositeLockPromise(int size) {
+         lockPromiseList = new ArrayList<>(size);
+         notifier = new CompletableFuture<>();
+      }
+
+      public void addLock(KeyAwareExtendedLockPromise lockPromise) {
+         lockPromiseList.add(lockPromise);
+      }
+
+      public void markListAsFinal() {
+         for (LockPromise lockPromise : lockPromiseList) {
+            lockPromise.addListener(this);
+         }
+      }
+
+      @Override
+      public boolean isAvailable() {
+         for (LockPromise lockPromise : lockPromiseList) {
+            if (!lockPromise.isAvailable()) {
+               return false;
+            }
+         }
+         return true;
+      }
+
+      @Override
+      public void lock() throws InterruptedException, TimeoutException {
+         Exception exception = null;
+         ExceptionType exceptionType = ExceptionType.NONE;
+         for (ExtendedLockPromise lockPromise : lockPromiseList) {
+            try {
+               //we still need to invoke lock in all the locks.
+               lockPromise.lock();
+            } catch (InterruptedException e) {
+               if (exception == null) {
+                  exception = e;
+                  exceptionType = ExceptionType.INTERRUPTED;
+               }
+            } catch (TimeoutException e) {
+               if (exception == null) {
+                  exception = e;
+                  exceptionType = ExceptionType.TIMEOUT;
+               }
+            } catch (DeadlockDetectedException e) {
+               if (exception == null) {
+                  exception = e;
+                  exceptionType = ExceptionType.DEADLOCK;
+               }
+            } catch (RuntimeException e) {
+               if (exception == null) {
+                  exception = e;
+                  exceptionType = ExceptionType.RUNTIME;
+               }
+            }
+         }
+         if (exception != null) {
+            switch (exceptionType) {
+               case INTERRUPTED:
+                  throw (InterruptedException) exception;
+               case TIMEOUT:
+                  throw (TimeoutException) exception;
+               case RUNTIME:
+                  throw (RuntimeException) exception;
+               case DEADLOCK:
+                  throw (DeadlockDetectedException) exception;
+               default:
+                  break;
+            }
+         }
+      }
+
+      @Override
+      public void addListener(LockListener listener) {
+         notifier.thenRun(() -> listener.onEvent(lockState));
+      }
+
+      @Override
+      public void setDeadlockChecker(DeadlockChecker deadlockChecker) {
+         for (LockPromise lockPromise : lockPromiseList) {
+            lockPromise.setDeadlockChecker(deadlockChecker);
+         }
+      }
+
+      @Override
+      public void onEvent(LockState state) {
+         if (state != LockState.AVAILABLE && UPDATER.compareAndSet(this, LockState.AVAILABLE, state)) {
```

Comment:
no. the ```lockState``` is only used to have information to trigger the listeners. 
and this method is invoked by all the locks in the list when they change state. so, if the composite receives a timeout or deadlock event, it will cancel all the locks.